[IMP] website, website_forum: redesign s_faq_collapse snippet

stefanorigano opened the pull request:
---
This pr/task is handled by @Brieuc-brd 

---


The aim of this commit is to improve the visual design of the accordion snippet and to introduce a new snippet using this accordion.

### 🌟 Introduce `s_accordion`  inner snippet
This commit introduces the new inner content `s_accordion` snippet.
- It offers three different models: default, borderless and boxed.
- The chevron can be set before or after the title.
- Each background item can be customized with color preset or custom   colors.

### Redesign `s_faq_collapse` snippet
This snippet has been redesigned to provide better rendering using the new `s_accordion` snippet.

task-3759048
Part of task-3619705

Requires: 
- https://github.com/odoo/design-themes/pull/820
- https://github.com/odoo/upgrade/pull/6309
---


### New `s_accordion`
| Before | After |
|--------|--------|
| ![Capture d’écran 2024-07-23 à 15 11 06](https://github.com/user-attachments/assets/5f7ddffb-58de-4546-bc7b-09d760e5446c) | ![Capture d’écran 2024-07-23 à 15 10 32](https://github.com/user-attachments/assets/e874cfce-cadb-4940-b13f-c65afb3c66a9) |
| ![Capture d’écran 2024-07-23 à 15 11 40](https://github.com/user-attachments/assets/6895c8c8-57ea-4391-8c4d-479c0ec2c26e) | ![Capture d’écran 2024-07-23 à 15 12 19](https://github.com/user-attachments/assets/850f450f-647b-46ef-9bfd-59f2f2c4a3ff) |
| ![Capture d’écran 2024-07-23 à 15 13 09](https://github.com/user-attachments/assets/9e5b4574-6487-4145-9d40-f545586b991e) | ![Capture d’écran 2024-07-23 à 15 13 25](https://github.com/user-attachments/assets/ec2ab7fd-3373-417d-a84d-a7c6d5da0610) |
| ![Capture d’écran 2024-07-23 à 15 13 47](https://github.com/user-attachments/assets/18490206-b7e4-4cd6-9813-0c0d5e513443) | Removed |

### `s_faq_collapse` redesign
| Before | After |
|--------|--------|
| ![Capture d’écran 2024-07-23 à 15 15 55](https://github.com/user-attachments/assets/965dc2e7-17cd-4ce7-9431-bc53edbfad68) | ![Capture d’écran 2024-07-23 à 15 15 25](https://github.com/user-attachments/assets/907f6045-1e51-4355-943f-d93dde1b22b7) |


---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/172577.png)](https://mergebot.odoo.com/odoo/odoo/pull/172577)

Brieuc-brd commented:
@stefanorigano 👋 
Ready for review 💪  ! Thanks in advance :)

stefanorigano commented on addons/website/views/snippets/s_accordion.xml:
```diff
@@ -0,0 +1,61 @@
+                    </div>
+                </div>
+            </div>
+            <div class="accordion-item position-relative z-1" data-name="Item" role="presentation">
```
Why is `z-1` needed? 🤔

Brieuc-brd replied:
Since the "card" template uses a pseudo-element as default background and has a negative z-index (which allows you to customize the card background color on top), this `z-1` ensures that this default background is always visible wherever you drag and drop the snippet.
Example when dropping the accordion in the footer :

| Without z-1 | With z-1 |
|--------|--------|
| <img width="1399" alt="Capture d’écran 2024-07-25 à 09 07 51" src="https://github.com/user-attachments/assets/39f1577d-d760-486f-aeac-0ff6a35ec0ff"> | <img width="1458" alt="Capture d’écran 2024-07-25 à 09 07 30" src="https://github.com/user-attachments/assets/33911c92-84b1-4754-9bd9-8de4ca6eb7ee"> |

stefanorigano commented on addons/website/views/snippets/s_accordion.xml:
```diff
@@ -0,0 +1,61 @@
+                    data-img="/website/static/src/img/snippets_options/pos_left.svg"/>
+                <we-button title="Right" data-select-class="justify-content-between"
+                    data-img="/website/static/src/img/snippets_options/pos_right.svg"/>
+            </we-button-group>
```
I'd add control over the border-radius, it definitely makes sense when the user chooses "Boxes" or apply a bg-color

stefanorigano commented on addons/website/views/snippets/s_accordion.xml:
```diff
@@ -0,0 +1,61 @@
+    <xpath expr="." position="inside">
+        <div data-selector=".s_accordion" data-target=".accordion">
+            <we-select string="Card Style">
+                <we-button title="Default" data-select-class="s_accordion_default">Default</we-button>
+                <we-button title="Borderless" data-select-class="accordion-flush">Borderless</we-button>
+                <we-button title="Boxed" data-select-class="s_accordion_boxed">Boxed</we-button>
```
I'm not sure about the UX, what about:

- Reword "_Default_" to "_Boxed_": default  it's not really explicative for the user while boxed it's closer to the final visual result
- Reword the "_Borderless_" to "_Side borders_": transform it to a switch, active by default and visible when the style is   "_Boxed_" (ex default) only.
- Reword "_Boxed_" to "_Highlight Active_":  it's what the option does at the end. Active elements become more predominant thanks to their background

<img width="262" alt="image" src="https://github.com/user-attachments/assets/694ae56a-c1c7-4a3b-965b-6ab1f20f2034">

stefanorigano commented on addons/website/views/snippets/snippets.xml:
```diff
@@ -600,12 +613,12 @@
 
-    <div data-js="MultipleItems" data-selector=".s_faq_collapse">
+    <div data-js="MultipleItems" data-selector=".s_accordion">
         <we-row string="Item">
```
ux
```suggestion
        <we-row string="Items">
```

stefanorigano commented on addons/website/views/snippets/snippets.xml:
```diff
@@ -600,12 +613,12 @@
         <we-row string="Item">
-            <we-button data-add-item="" data-item=".accordion > .card:last" data-select-item="" data-no-preview="true" class="o_we_bg_brand_primary">
+            <we-button data-add-item="" data-item=".accordion > .accordion-item:last" data-select-item="" data-no-preview="true" class="o_we_bg_brand_primary">
                 Add Item
```
ux
```suggestion
                Add New
```

stefanorigano commented on addons/website/views/snippets/s_accordion.xml:
```diff
@@ -0,0 +1,61 @@
+                <we-button title="Borderless" data-select-class="accordion-flush">Borderless</we-button>
+                <we-button title="Boxed" data-select-class="s_accordion_boxed">Boxed</we-button>
+            </we-select>
+            <we-button-group string="Collapse Icon" data-apply-to=".accordion-header">
```
ux
```suggestion
            <we-button-group string="Icon Position" data-apply-to=".accordion-header">
```

stefanorigano commented on addons/website/static/src/snippets/s_faq_collapse/000.xml:
```diff
@@ -0,0 +1,64 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
```
Not sure to understand what you're doing here  🤔

`00X.xml` files into the `static` folder are supposed to contain QWeb JS templates, loaded in JS, eg https://github.com/odoo/odoo/blob/master/addons/website/static/src/snippets/s_countdown/000.xml

Brieuc-brd replied:
As the template has totally changed, I thought I should keep a backup. I removed it :)

stefanorigano reviewed (commented):
🔴 Thanks @Brieuc-brd , yet another "simple" one, right ;)
Some comments for you... and one big question about `addons/website/static/src/snippets/s_faq_collapse/000.xml` 🤔

Brieuc-brd commented:
Hey @stefanorigano 👋 
I've updated the branches, could you have a look ?
Thanks in advance !

stefanorigano reviewed (commented):
Hello @Brieuc-brd , I've just pushed 2commits on top

3323b8311d798c1fecb196a7fed212dd86b319f6

- increase pseudo element opacity
- hide border-radius controller if no color is set
- move background color controller to the main wrapper: it's uncommon to choose different colors for each entry.  Having the background option located into another section was confusing.

| before  | after |
|--------|--------|
| ![image](https://github.com/user-attachments/assets/62806838-e886-45c9-887b-0dd147c3af00) | <img width="267" alt="image" src="https://github.com/user-attachments/assets/088e7f3b-bbf1-4ec5-abee-7991a30ebd40"> <img width="267" alt="image" src="https://github.com/user-attachments/assets/2f07bffb-50ee-41c6-a88d-1f0deafe6f3a">| 

9c25c62d62f282fb4bfa854d4714dc5eb5efc86c

- Streamlines even more the UI moving the accordion's options into its parent column (faq snippet only).

| commit 1 | this commit |
|--------|--------|
| <img width="268" alt="image" src="https://github.com/user-attachments/assets/148ed708-2e13-487e-a64a-1af60a6e46cc"> | <img width="268" alt="image" src="https://github.com/user-attachments/assets/1eb617d6-8d05-4e4e-a011-24b88e39d148"> | 

This last commit is more "experimental" and will probably need some adaptation, probably with the help of a dev. At the moment the downside is that empty section are still rendered but I'm sure that there is a easy way to hide them.

All branches rebased 👍 

I'll let you check, thanks!

Brieuc-brd commented:
Hello @stefanorigano, thanks for your changes.

Changing the level of the color feature breaks the "Highlight Active" template, we will have to take a look to fix that:
![Capture d’écran 2024-07-26 à 08 16 45](https://github.com/user-attachments/assets/23bc4012-99c1-4df6-93ca-74ffc6a5679a)

Also, I'm not convinced about hiding the Round corners feature if the user doesn't change the background color:
- If the user chooses the “Highlight Active” template and keeps the default background color, the rounded corners cannot be edited. The user will have to select approximately the same color as the default to unlock the rounded corners function.
- If the user wants to add a classic accordion that matches the background of the current snippet, he will need to select the same background color to unlock the rounded corners. IMO this is not the expected behavior 😬 .

| Feature Locked | Feature Unlocked |
|--------|--------|
| ![Capture d’écran 2024-07-26 à 08 15 54](https://github.com/user-attachments/assets/5d74294f-eb98-4ea2-b030-32401d83a762) | ![Capture d’écran 2024-07-26 à 08 22 38](https://github.com/user-attachments/assets/f0ee0379-38b5-4563-9296-d792145da01c) |

Brieuc-brd reviewed (changes requested):
Hello @stefanorigano, thanks for your changes.

Changing the level of the color feature breaks the "Highlight Active" template, we will have to take a look to fix that:
![Capture d’écran 2024-07-26 à 08 16 45](https://github.com/user-attachments/assets/23bc4012-99c1-4df6-93ca-74ffc6a5679a)

Also, I'm not convinced about hiding the Round corners feature if the user doesn't change the background color:
- If the user chooses the “Highlight Active” template and keeps the default background color, the rounded corners cannot be edited. The user will have to select approximately the same color as the default to unlock the rounded corners function.
- If the user wants to add a classic accordion that matches the background of the current snippet, he will need to select the same background color to unlock the rounded corners. IMO this is not the expected behavior 😬 .

| Feature Locked | Feature Unlocked |
|--------|--------|
| ![Capture d’écran 2024-07-26 à 08 15 54](https://github.com/user-attachments/assets/5d74294f-eb98-4ea2-b030-32401d83a762) | ![Capture d’écran 2024-07-26 à 08 22 38](https://github.com/user-attachments/assets/f0ee0379-38b5-4563-9296-d792145da01c) |

stefanorigano commented:
Hello @Brieuc-brd , 
- https://github.com/odoo/odoo/pull/172577/commits/3323b8311d798c1fecb196a7fed212dd86b319f6: fix missing condition
- https://github.com/odoo/odoo/pull/172577/commits/9c25c62d62f282fb4bfa854d4714dc5eb5efc86c: simplify logic by avoiding to use `data-exclude`

diff: https://github.com/odoo/odoo/compare/ff13664cedb984da2046c236eadea8cb31cbdee4..9c25c62d62f282fb4bfa854d4714dc5eb5efc86c

stefanorigano commented:
> Do you have an idea how could we fix that ?

Nope, I'm sorry :/

> you might want an accordion "boxed" that fit with the background of the parent snippet.

Right, I missed this condition, it should be fine now :)

> I still think that the round corner feature should be always visible. The only issue I see is when the user would change the round corners for a "side border" accordion, without background color: in this case nothing will happens.
But I think this is the excepted behavior. Once the user will change the template or add a background color he will see the corner radius. To me this is fine 🙂

No, don't agree. We should avoid showing option that do nothing... for users that don't know CSS it's not even the  "expected behavior" IMO. For them if the option is there should perform some sort of action... and they will try it.
Also, this is a "mixed section" that target different stuff, the wrapper , the items, the icons... even pro-users may get  confused a bit if they miss the dependencies graphical indicators.

It's the same reason why we hide border options when there is no border... when possible let's do it, it just make sense ;)
Moreover this is really an easy fix, it should be done now.

| no border | border |
|--------|--------|
| <img width="277" alt="image" src="https://github.com/user-attachments/assets/34abde16-37e0-40d9-81e4-2d876517020e"> | <img width="281" alt="image" src="https://github.com/user-attachments/assets/e3909444-2bc2-4d97-8b89-a3f9fa57a633"> |

Brieuc-brd commented:
Hey @stefanorigano 👋 I've updated the branch 👍 

> Right, I missed this condition, it should be fine now :)

About the border feature, I saw your changes and you were right, it’s better as you did, thanks 👍

> Nope, I'm sorry :/

I found a way to prevent the `!important` tag on colors, thanks to the `data-prevent-important="true"`  attribute. 

But there’s still a bug when the color picker is in the “Accordion block”: it seems to recognize only the first accordion item. The color picker can bug depending on which accordion item you click on.

I haven’t found a solution for that, so I moved back the color picker for each accordion item, which works fine 👍 .
This allows the user to create more personalized accordions using different colors, which is a great feature when you want to visually organize your content:

![Capture d’écran 2024-07-29 à 11 01 36](https://github.com/user-attachments/assets/c6bf9b3c-9106-4e2b-8d50-20aab64622f1)

What do you think ?

stefanorigano commented:
Hello @Brieuc-brd , thanks for your efforts here.

> I haven’t found a solution for that, so I moved back the color picker for each accordion item, which works fine 👍 .

Unfortunately it doesn't. The JS that toggles the border-radius controller visibility now fails, unless a color is applied to the very first entry [1] 🙁
Also, depending by what you click first, when "Highlight active" design is active the the bg-color is not always correctly recognized  by the editor [2] .
For this last point it was already the case even if,  reading your message, I hoped that the last changes could have solve it somehow... but they don't 😕
The issue it's indeed related to enforced transparency we applied on collapsed element, since the editor it's not really made to apply colors according to the element's "state" ("not collapsed" in this case).

We can always improve this later and for the sake of merging this ASAP I propose to remove the JS and make the border-radius controller always visible (🥲). The controller will look "broken" in some circumstances, but it this would solve "[1]".

"[2]" cannot be solved without making use of huge JS black magic, but the little UX issue it's probably a good trade if we want to save this design.
Yet again, since **empty** sidebar sections cannot be removed apparently (and for the sake or speeding up tech. review), let's remove the data-selector optimization I've introduced (🥲🥲).
As result we'll get a lot of option for the column that are not necessary, but having the color controller into "Accordion Item" section will look less weird this way. Before applying this changes let's fork a backup branch tho, I'm sure that we'll be ask to address the edition complexity of this component later on 👍 

Thanks again

| [1] | [2] |
|--------|--------|
| ![image](https://github.com/user-attachments/assets/a762ff63-6475-4718-951c-8ecc9e578934) | ![image](https://github.com/user-attachments/assets/e6d80354-740f-4557-8584-7415f18c4a23) |

Brieuc-brd commented:
Ok 👍 
I've made the changes, here's the diff to be sure I didn't miss anything :
https://github.com/odoo/odoo/compare/9044b08e6d7defe011ffa3a805b8d56141f9318e..2cc3b227f9b931af0db417bdba0b315e96e7e7e6

I also made a backup of this branch before applying the changes :)

stefanorigano commented:
Hello @Brieuc-brd , thanks.
Let's move this forward 👍

robinlej reviewed (commented):
Hello @Brieuc-brd, here are my comments. It generally works well and I don't have deeply technical blocking points. :+1: My main point of concern is the accessibility.

robinlej commented:
Not really tech review, but I think there's something a bit weird with the "Highlight Active" mode: say your page background is white/transparent, if you want your accordion items to be black with white text (or any dark background with clear text, you get the point) when you open them, that can't work because you can't set a different color for the button when it's closed (so the clear text isn't readable when the accordion is closed). Couldn't we set 2 colorpickers, one for when it's closed (default to transparent) and one for when it's open?

More generally, any dark background on the accordion will be a problem for the arrow, as the arrow color relies on the preset and can't be changed manually, and the preset colors can't be put on the accordion itself but need to be on the ancestor snippet. Shouldn't it be possible on the accordion?

Brieuc-brd commented on addons/website/static/src/snippets/s_accordion/000.scss:
```diff
@@ -0,0 +1,69 @@
+    .s_accordion_highlight {
+        .accordion-item {
+            border: 0;
+            border-radius: var(--#{$prefix}accordion-border-radius);
+            overflow: hidden;
```
The `overflow: hidden` is necessary to clip the entire item when you set a custom background color. Otherwise there would be a rounded corner issue:

| Without overflow | With overflo |
|--------|--------|
| <img width="805" alt="Capture d’écran 2024-08-08 à 14 12 27" src="https://github.com/user-attachments/assets/ad678d7e-6a93-4c3e-a81c-5ac73406f703"> | <img width="820" alt="Capture d’écran 2024-08-08 à 14 13 16" src="https://github.com/user-attachments/assets/612c90c1-bbc0-4729-908e-3f9f3e07eb81"> |

Brieuc-brd replied:
I've updated the branch to improve the focus state ;)

Brieuc-brd commented on addons/website/static/src/snippets/s_accordion/000.scss:
```diff
@@ -0,0 +1,69 @@
+            overflow: hidden;
+
+            // This removes the bg color when the item isn't displayed.
+            &:has(.collapsed) {
```
I haven't found any other way to do it. 

Since this pseudo-class is more than 92% covered, I think we can use it.
https://caniuse.com/?search=%3Ahas()

Also, in this case, `has()` is defined for decorative purpose (it allows only the background of the active element to be displayed). This means that even in the worst scenario when the user's browser can't render this selector, he can still see the content without having any readability issue.


| rendering with`has()`✅ | rendering without `has()`❌ |
|--------|--------|
| <img width="1403" alt="Capture d’écran 2024-08-08 à 14 59 26" src="https://github.com/user-attachments/assets/6de50257-1e66-4fbe-99b7-5ee6c1d96795"> | <img width="1374" alt="Capture d’écran 2024-08-08 à 15 02 47" src="https://github.com/user-attachments/assets/432f658f-c7c7-4b8b-bb97-020940eb2a9a"> |

robinlej replied:
Yes, I forgot to update my comment, but indeed it's not blocking for browsers that don't support it. :+1: That's good for me

Brieuc-brd commented on addons/website_sale/static/src/scss/website_sale.scss:
```diff
@@ -814,6 +814,6 @@ a.no-decoration {
 
-.accordion-button:not(.collapsed).bg-transparent::after {
+.wsale_products_categories_list .accordion-button:not(.collapsed).bg-transparent::after {
     background-image: escape-svg($accordion-button-icon);
```
> that rule doesn't work (the icon doesn't show)

In fact, this icon is displayed behind the UI icon as a background image  😬 👇 
<img width="794" alt="Capture d’écran 2024-08-08 à 16 58 30" src="https://github.com/user-attachments/assets/88a0b497-89e1-45e3-a4b4-a10a390b0a04">

This adaptation allows to scope only the related module and avoids to break other accordions.

robinlej replied:
Sorry, my message wasn't clear. I understood the reason behind your fix, I'm just pointing out that their rule doesn't work in the first place. Unless I'm completely wrong, it's supposed to be there when certain options are toggled on:
![image](https://github.com/user-attachments/assets/2f19d5f7-d529-4f99-a277-eb9571c3098f)

But since the pseudo-element doesn't have any width/height, it's not shown.

So my comment wasn't about this PR, just a way to point the problem. But I should have done that in their channel. ;)

Brieuc-brd commented on addons/website/views/snippets/snippets.xml:
```diff
@@ -116,7 +116,7 @@
                     <keywords>customers, clients</keywords>
                 </t>
-                <t t-snippet="website.s_quotes_carousel" string="Quotes" group="people">
+                <t t-snippet="website.s_quotes_carousel" string="FAQ" group="people">
```
I've updated the content 👍  Now it's a real FAQ 😄

robinlej replied:
I mean this is the testimonials carousel, not the accordion FAQ (which you changed and is just below, no problem with that!). Am I missing something obvious? :sweat_smile:

Brieuc-brd replied:
Oh my !, too much extra miles started to burn my brain 🤦 
Sorry for that 😅, I made the change : https://github.com/odoo/odoo/compare/271081a805ccce5011105a1ebc50237945cdf043..47dfa2c41574df5f0e02a2ebd92acd821af1dfb1

But I'll keep the new texts, it's better tho :)

Brieuc-brd commented:
Hey @robinlej 👋 
Thanks for your review 👍 
I've updated the branch to fix all accessibility and content issues.
I've just pushed it, this should be green, but I keep an eye on it if it turns red.

The diffs :
- https://github.com/odoo/odoo/compare/e456d9b54b45304775a76b706f29a6ffedfd11d6..0807dfd5e9242b640c83220754e18f795d0990a4
- https://github.com/odoo/odoo/compare/f37d7c54b89f2a7476a61eb50d41128549ead08a..271081a805ccce5011105a1ebc50237945cdf043

Could you take a look ?
Thanks in advance !

robinlej commented:
Just force-pushed an update for the JS method to do a small fix, use vanilla JS instead of jQuery, and rename the variables to be consistent with the new structure.
It's all good for me, thanks for the work! :+1:

Brieuc-brd commented:
@robinlej 👋 
Little change (https://github.com/odoo/odoo/compare/1495b8817330cdc03902df723ac28faaf6dcb050..5b8b1693c6c071236e86abc8a3ff414b07e7d643) + rebase 👍

sobo-odoo reviewed (approved):
ROLE said that it was good to go, so let's r+ ! :+1: 

@robodoo r+

robodoo commented:
@stefanorigano @sobo-odoo linked pull request(s) odoo/upgrade#6309 not ready. Linked PRs are not staged until all of them are ready.

sobo-odoo commented:
@robodoo r+

Brieuc-brd commented:
@sobo-odoo 👋 
There was a conflict in the branch 🥲 I had to rebase it, should you r+ again?

sobo-odoo commented:
@robodoo r+
